### ISSUE TITLE
Inputing also ".rst" files

### ISSUE THREAD

**Issue by GaelVaroquaux** [sphinx-gallery/sphinx-gallery#330]
If we ever want to also input ".rst" files, it's really easy to read them in, and convert them to .py files.

I just came across some code lying on my disk since a while (probably copy-pasted from Internet):

```python
from docutils.core import publish_doctree

doctree = publish_doctree(open('input.rst').read())

def is_code_block(node):
    return (node.tagname == 'literal_block')

code_blocks = doctree.traverse(condition=is_code_block)
source_code = [block.astext() for block in code_blocks]

with open('output.py', 'w') as f:
    f.write('\n######################\n\n'.join(source_code[2:]))
```

So we should keep in mind that it's probably a low cost feature that might make users happy.

**Comment by ezyang** [comment]
Seconding this; pytorch/tutorials has this line: https://github.com/pytorch/tutorials/blob/67032d771a93fa996deadfdc4d5c7e614ab4c16d/conf.py#L81 to get files from the source directory to the gallery directory, so that we can have documentation that intermixes galleries with manually written files. It would be much better if we did not have to do this ourselves.

**Comment by larsoner** [comment]
IIUC, these two seem a bit different:

- In @GaelVaroquaux's case, the RST is converted to `.py` so it gets run by Sphinx-gallery, and then converted back to RST (with all figures, backrefs, doc links, etc.) *without any plain RST in between* (though you could probably enhance the code to take care of this limitation)
- In @ezyang's case, RST is just copied directly, so no figures/backrefs/doc links are created by any literal blocks

@ezyang do you actually have RST with literal blocks that need to be run? I guess what would satisfy both use cases would be allowing RST with literal blocks and executing (and embedding figures and code links) for any literal blocks while preserving the rest of the RST.

Do you agree that would work for both of your use cases @GaelVaroquaux @ezyang ? If so, then I don't see a problem with adding this feature. I think for most repos it wouldn't make much difference, except that we might permit projects to use `README.rst` [in addition to](https://github.com/sphinx-gallery/sphinx-gallery/blob/853cc02e27bcc6d0f88f5ca0601ccac94766f39d/sphinx_gallery/gen_rst.py#L313) the more standard `README.txt`. But I guess any repos with this pattern could just `git mv README.rst README.txt` and then they would be okay. To avoid breaking things for them, though, we might want to have a `run_rst=False` by default or so, and PyTorch could just set `run_rst=True`.

**Comment by lucyleeow** [comment]
> except that we might permit projects to use README.rst in addition to the more standard README.txt

I thought we already allowed both `README.txt` and `README.rst` ? I think `config['source_suffix']` is `.rst` ? Maybe I am confused as to what you are saying?

**Comment by larsoner** [comment]
Yes we already do it -- I'm saying that this might be a problem / conflict if we implement a "run all RST" feature

**Comment by lucyleeow** [comment]
Ah I get you, thanks!

**Comment by ezyang** [comment]
> @ezyang do you actually have RST with literal blocks that need to be run?

Not actually. I think it's just that we have to store both Gallery py files and non-Gallery rst files in the same "source" directory, and go into the same destination (being in the same destination is important because the ultimate directory hierarchy of the final directory is indifferent to whether or not a particular document was Gallery py or just plain rst. So maybe I should not have chimed in on this particular issue. :)

**Comment by bdice** [comment]
I _think_ I want the same thing as @GaelVaroquaux but I'm not sure - I'm new to sphinx-gallery and am trying to assess whether to use it for our docs. I currently have rST files and I really want to get sphinx-gallery's automatic intersphinx references inserted into all code snippets (doctest-style snippets in method docstrings and rST blocks like `.. code-block:: python`. Is that all handled inside of [docs_resolv.py](https://github.com/sphinx-gallery/sphinx-gallery/blob/master/sphinx_gallery/docs_resolv.py)? If there is another way someone might know just to get the intersphinx link-insertion feature, please let me know! Here's an [example file](https://github.com/glotzerlab/signac-docs/blob/master/docs/source/recipes.rst).

**Comment by larsoner** [comment]
> I currently have rST files and I really want to get sphinx-gallery's automatic intersphinx references inserted into all code snippets (doctest-style snippets in method docstrings and rST blocks like .. code-block:: python). Is that all handled inside of docs_resolv.py? If there is another way someone might know just to get the intersphinx link-insertion feature, please let me know! 

The code links are [mostly generated by live variable class inspection during execution](https://github.com/sphinx-gallery/sphinx-gallery/blob/master/sphinx_gallery/backreferences.py#L82-L142). In principle you might be able to get something similar kind of working by inspecting the return types of all functions, but that's going to be outside the scope of Sphinx-gallery (and probably a lot of work). In other words, you probably need sphinx-gallery to execute your code to insert references to classes, methods, and functions.

Is there some reason you don't want to convert your docs to `.py` files that Sphinx-Gallery can execute for you? This is really a (the?) core idea behind Sphinx-Gallery -- it executes scripts and HTML-ifies outputs that your users like would see if they executed the scripts themselves.

**Comment by bdice** [comment]
@larsoner Thanks for the info! I was trying to see how much intersphinx magic I could accomplish without rewriting/reformatting docs (perhaps only adding a plugin with a single line of code). 😅 I have looked at the demos, and this package makes perfect sense for "example galleries." However, I think my primary question is a little different (or I'm not understanding). This might be more suited for a separate issue, or could be out of scope.

I want to get [user API documentation with Example snippets like this one](https://freud.readthedocs.io/en/latest/modules/box.html#freud.box.Box.center_of_mass) to show the same kind of the intersphinx linking that sphinx-gallery provides for notebooks. The snippets are executable but it doesn't seem like sphinx-gallery would work with that since API docstrings are not really a notebook-like format. I posted in this issue instead of creating a new issue because I hoped that adding an rST input format might make it possible to intersphinx-ify the examples in API docs (when passing through sphinx-autodoc).

For example, scikit-learn has [API doc examples](https://scikit-learn.org/stable/modules/generated/sklearn.cluster.KMeans.html#sklearn-cluster-kmeans) (where the intersphinx links are _missing_), and a [standalone "example" notebook](https://scikit-learn.org/stable/auto_examples/cluster/plot_cluster_iris.html#sphx-glr-auto-examples-cluster-plot-cluster-iris-py) that uses sphinx-gallery (which adds intersphinx links).

**Comment by larsoner** [comment]
> The snippets are executable but it doesn't seem like sphinx-gallery would work with that since API docstrings are not really a notebook-like format.

Sphinx-Gallery does not actually execute notebooks. It executes standard Python scripts, then also happens to output `ipynb` (and `py` and `rst`) files that correspond to them. So actually it seems like it might be tractable.

>  (by executing them when passing through the rST generated by sphinx-autodoc?)

This is where I think it will get complicated. For Sphinx-Gallery to do linking, it needs to actually execute the code itself and inspect the variables. Who executes the code in the example above in your doc build? If the answer is "currently nobody during the Sphinx build, they are only run for example when doing pytest --with-doctest or something similar" then in principle there might be a way to get Sphinx-Gallery to:

1. hook into the autodoc/numpydoc/whatever build process
2. recognize that there is example code
3. execute it
4. figure out what to link where, and
5. insert links in the right places.

1/2/5 will be the hard parts, as 3 and 4 are pretty straightforward refactorings of existing functions. 5 might be tricky because we currently modify the built HTML output files rather than insert refs in the RST, and this seems like a fragile method when you're talking about an autodoc-generated HTML file (which could be quite long with *much* more extensive variable name reuse, causing likely errant substitutions).

To figure out how to do 1 and 2, I would look at the code for matplotlib's `.. plot::` directive, and also `numpydoc_use_plots` (which automatically inserts `.. plot::` lines). This will tell you the sphinx mechanics necessary to get a callable that could then call the right Sphinx-Gallery functions to execute the given code block(s) and produce the outputs and linking bits.

I don't think this is a trivial exercise but I could see this being valuable for existing users of Sphinx-Gallery, so I think it might actually be in scope. It's also complicated by the fact that there might be Sphinx-Gallery users already using `numpydoc_use_plots`, so our new behavior would need to be opt-in, and we'd want to detect the conflicting `numpydoc_use_plots` and throw an error saying that there is a conflict.

@bdice would you be interested in trying to implement some or all of this?

@GaelVaroquaux @lucyleeow do you see this as being useful for sklearn as well? As @bdice points out, the `Examples` code for sklearn classes does not get the nice linking that sklearn example scripts do. If so, any interest in working on this as well @lucyleeow ?

**Comment by lucyleeow** [comment]
> I don't think this is a trivial exercise

Agreed!
I am not very familiar with numpydoc or autodoc. For step 1 - would the aim to be make it work for one, or the most common 'doc' used or if it works for one it will work for the others?

This does look interesting but very complicated. At the moment I am in the middle of moving countries (x1.5 - it's complicated) so I won't have time in the near future to do this. If someone else takes this on that would be great. If not I can take a look later.

**Comment by bdice** [comment]
@larsoner Excellent response! This helped explain a lot of the parts that I didn't know. I am interested in pursuing this concept further, especially if there is interest from other sphinx-gallery users. I'm probably going to work on enabling sphinx-gallery for a project first, so I can get more familiar with its usage and features.

@lucyleeow Good question. I use standard Sphinx-style, NumPy-style, and Google-style docstrings in different projects, so I would want to support all of them, but I think that should be possible. I would guess that there's a layer/stage of the process at which all example code snippets are handled the same way and could be detected/executed without needing to specialize for different formats/tools, but I am not super familiar with the internals of these packages.
